### PR TITLE
Item Versioning

### LINKED ISSUE #2844: Create new Item Version
From release plan spreadsheet

Estimate from release plan: 72 hours
Expressing interest: none

https://wiki.duraspace.org/display/DSDOC6x/Item+Level+Versioning  In JSPUI, possible for submitters to version own items. In XMLUI, only Admins or Com/Col Admins can version.

### LINKED ISSUE #2847: Create a new item version as submitter
From release plan spreadsheet

Estimate from release plan: 10 hours
Expressing interest: none

This feature only existed in JSPUI

### PR DESCRIPTION
**Pull request by Micheleboychuk**
## Description
Rest Contract: https://github.com/DSpace/RestContract/pull/171
Fixes #2844
Fixes #2847
Frontend PR: https://github.com/DSpace/dspace-angular/pull/1318

This PR relates to the versioning features in DSpace 7. This first PR only deal with normal items without making any explicit decision about relationship (i.e. they are ignored as by current Version*Service implementation.
In this PR we introduce the endpoints needed to
- show a warning message in the item page if it is not the latest version with a link to the current version
- access to the full version history of an item (if any). This will be needed in the public item page to allow navigation and in the edit screen to allow management
- create a new version from a previous one

The following authorization features are introduced
- CanEditVersion
- CanDeleteVersion
- CanCreateVersion

Please note that by REST Contract delete of a version is delegated to the normal items endpoint

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by abollini** [comment]
@tdonohue @benbosman please don't spent time on reviewing this pr as we found that the rest contract and the implementation needs further discussion

**Comment by tdonohue** [comment]
@Micheleboychuk : in case you haven't noticed them, I wanted to call out the LGTM comments above...it's saying that you've accidentally created 3 new `Spurious Javadoc @param tags` issues in the codebase.  I realize this is a work in progress still, but obviously it'd be good to clean up any LGTM issues as you finish this up. Thanks!

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionHistoryDraftVersionLinkRepository.java
Comment (KevinVdV) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionHistoryDraftVersionLinkRepository.java:60]: Typo in the "provided" word.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionHistoryDraftVersionLinkRepository.java
Comment (KevinVdV) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionHistoryDraftVersionLinkRepository.java:64]: Typo in the version word.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionRestRepository.java
Comment (KevinVdV) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionRestRepository.java:116]: Same as above.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionRestRepository.java
Comment (KevinVdV) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionRestRepository.java:143]: Perhaps add a string here explaining why this exception is thrown ?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionRestRepository.java
Comment (KevinVdV) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionRestRepository.java:171]: Perhaps rephrase into something like:
The 'summary' already contains the value {value}, it is not possible to add a new value.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/security/ExtractorOfAInprogressSubmissionInformations.java
Comment (KevinVdV) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/security/ExtractorOfAInprogressSubmissionInformations.java:39]: Class could use a bit of documentation on how to use.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionHistoryRestRepositoryIT.java
Comment (KevinVdV) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionHistoryRestRepositoryIT.java:415]: Typo in the word deleting

**Review by KevinVdV — CHANGES_REQUESTED** [review]
Made an initial passthrough of the code review. Things are looking good already, see the minor comments.

In addition to the comments I couldn't find tests for the following (correct me if they are there but I missed them):
* api/versioning/versionhistories/<:versionHistoryId>/draftVersion
     *  Test with an non existing version history
     * Test with an existing version history but NO draft version
     * Test with a workflow item as a "draft version"
* POST /api/versioning/versions[?summary=<summary-text>]
    * Test without a summary
    * Test for creating a version of a workflow item (this should fail ofc)

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionRestRepository.java
Comment (mwoodiupui) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/VersionRestRepository.java:125]: This looks like two distinct problems ORed together.  The first looks like authorization (and I agree that AuthorizeException seems appropriate), while the second seems to be that entities are forbidden here?  I suggest separate tests with separate throws, and that each exception should state its objection to the call.

**Comment by Micheleboychuk** [comment]
@KevinVdV thanks for your feedback, I have added all the changes, 
the lasted required test is already present
- Test for creating a version of a workflow item (this should fail ofc)
https://github.com/DSpace/DSpace/blob/496bb35eea6ce94c75b4777359e1520cb09c66a1/dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java#L561

**Review by KevinVdV — APPROVED** [review]
Except for two small "cleanup" remarks everything appears in order for me.
This PR can be merged once those small things have been done.

**Review by KevinVdV — APPROVED** [review]
My latest minor grievances were resolved, approving !

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionHistoryRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionHistoryRestRepositoryIT.java:206]: Shouldn't we be checking that these versions are in a very *specific* order instead of "AnyOrder"?  I believe versions are ordered by ID.  So, this should likely be `Matchers.containsInRelativeOrder(`

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionHistoryRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionHistoryRestRepositoryIT.java:239]: Again, you can remove this line, it resets automatically to 'false'

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java:508]: This doesn't appear to be deleting the version you just POSTed.  It looks like it's deleting the version you created via VersionBuilder, unless I'm misreading the code.  I think this last POST needs to have an `andDo` clause to update the idRef.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java:657]: Again, can remove this line as it's unnecessary

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java:721]: Can remove this line, it's unnecessary

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java:766]: Again, can remove

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/VersionRestRepositoryIT.java:813]: Again, can remove these two lines

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java:148]: You don't need this line, as the value resets after every test.  Just change the value at the beginning of a test if you need to do so

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java:218]: Again, can remove this line

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java:355]: Again, can remove this line

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java:404]: Again, can remove

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanCreateVersionFeatureIT.java:454]: Again, can remove these two lines

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanDeleteVersionFeatureIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanDeleteVersionFeatureIT.java:360]: Again, can remove

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanEditVersionFeatureIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanEditVersionFeatureIT.java:347]: Again, can remove

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanManageVersionsFeatureIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/authorization/CanManageVersionsFeatureIT.java:387]: Again, can remove

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Micheleboychuk : Apologies for the late review.  The code mostly looks good and appears to work in my initial testing.  However, there's a lot of (very minor) things that can be cleaned up in your tests primarily (and some requests for JavaDocs).  Overall though, good job.

File: dspace/config/modules/versioning.cfg
Comment (tdonohue) [inline dspace/config/modules/versioning.cfg:8]: Tiny thing, usually we put the default value after the `=` sign.  So, we might want to make this say `# versioning.enabled=true`

File: dspace/config/modules/versioning.cfg
Comment (tdonohue) [inline dspace/config/modules/versioning.cfg:26]: Tiny thing, usually we put the default value after the `=` sign.  So, we might want to make this say `# versioning.block.entity=true`

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Micheleboychuk :  I re-reviewed & re-tested this today.  First off, the code looks good now...thanks for the changes.  I only have a few very minor suggestions below.

In terms of testing, I found a bug when testing this with the UI which I've described in the UI ticket feedback here: https://github.com/DSpace/dspace-angular/pull/1318#pullrequestreview-768373105   I'm not sure if the bug is on the frontend or backend, but I'm hoping that you and @davidenegretti-4science will be able to track it down based on my instructions for how to reproduce it.

Once that bug is resolved, I think both of these PRs are looking good.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @Micheleboychuk ! This all looks good to me now.  I also tested it with the frontend again today and tried versioning off works (`versioning.enabled=false`) and restricting the version history works (`versioning.item.history.view.admin=true`), and those both work seamlessly as well.

So, I think this is ready to be merged.  That said, I'm going to wait on any final testing feedback on the UI PR prior to merging.

**Review by mwoodiupui — APPROVED** [review]
I reviewed only the code on which I had commented.  The changes look correct.

**Pull request merged**
